lib/elixir/lib/kernel.ex

```diff
@@ -1972,20 +1972,20 @@ defmodule Kernel do
   and the deletion will be considered a success.
   """
   @spec pop_in(Access.t, nonempty_list(term)) :: {term, Access.t}
-  def pop_in(data, keys)
-  def pop_in(nil, [h | _]), do: Access.pop(nil, h)
-  def pop_in(data, keys), do: do_pop_in(data, keys)
+  def pop_in(structure, keys)
+  def pop_in(nil, [key | _]), do: Access.pop(nil, key)
+  def pop_in(structure, keys) when is_list(keys), do: pop_in_structure(structure, keys)
 
-  defp do_pop_in(nil, [_ | _]),
+  defp pop_in_structure(nil, [_ | _]),
     do: :pop
-  defp do_pop_in(data, [h]) when is_function(h),
-    do: h.(:get_and_update, data, fn _ -> :pop end)
-  defp do_pop_in(data, [h | t]) when is_function(h),
-    do: h.(:get_and_update, data, &do_pop_in(&1, t))
-  defp do_pop_in(data, [h]),
-    do: Access.pop(data, h)
-  defp do_pop_in(data, [h | t]),
-    do: Access.get_and_update(data, h, &do_pop_in(&1, t))
+  defp pop_in_structure(structure, [fun]) when is_function(fun, 3),
+    do: fun.(:get_and_update, structure, fn _ -> :pop end)
+  defp pop_in_structure(structure, [fun | tail]) when is_function(fun, 3),
```

Comment:
We agreed to not use the `is_function/2` guard because the arity error invoked by using the wrong arity is more precise.